packages/embedder/embedder.ts

```diff
@@ -0,0 +1,250 @@
+})
+
+const refreshMetadata = async () => {
+  await refreshRetroDiscussionTopicsMeta(dataLoader)
+  // In the future, other sorts of objects to index could be added here...
+}
+const maybeQueueMetadataItems = async (modelManager: ModelManager) => {
+  const queueLength = await redisClient.zcard('embedder:queue')
+  if (queueLength >= Q_MAX_LENGTH) return
+  const itemCountToQueue = Q_MAX_LENGTH - queueLength
+  const configuredModels = [...modelManager.getEmbeddingModelsIter()].map((m) => m.getTableName())
+  const orgIds = await orgIdsWithFeatureFlag()
+
+  // For each configured embedding model, select rows from EmbeddingsMetadata
+  // that haven't been calculated nor exist in the EmbeddingsJobQueue yet
+  //
+  // Notes:
+  //   * `em.models @> ARRAY[v.model]` is an indexed query
+  //   * I don't love all overrides, I wish there was a better way
+  //     see: https://github.com/kysely-org/kysely/issues/872
+
+  const batchToQueue = await selectMetaToQueue(configuredModels, orgIds, itemCountToQueue)
+
+  if (!batchToQueue.length) {
+    console.log(`embedder: no new items to queue`)
+    return
+  }
+
+  const ejqHash: {
+    [key: string]: {
+      refUpdatedAt: Date
+    }
+  } = {}
+  const makeKey = (item: {objectType: string | null; refId: string | null}) =>
+    `${item.objectType}:${item.refId}`
+
+  const ejqValues = batchToQueue.map((item) => {
+    ejqHash[makeKey(item)] = {
+      refUpdatedAt: item.refUpdatedAt
+    }
+    return {
+      objectType: item.objectType,
+      refId: item.refId as string,
+      model: item.model,
+      state: 'queued' as DBInsert['EmbeddingsJobQueue']['state']
+    }
+  })
+
+  const ejqRows = await insertNewJobs(ejqValues)
+
+  ejqRows.forEach((item) => {
+    const {refUpdatedAt} = ejqHash[makeKey(item)]!
+    const score = new Date(refUpdatedAt).getTime()
+    redisClient.zadd('embedder:queue', score, item.id)
+  })
+
+  console.log(`embedder: queued ${batchToQueue.length} items`)
+}
+
+const dequeueAndEmbedUntilEmpty = async (modelManager: ModelManager) => {
+  while (true) {
+    const maybeRedisQItem = await redisClient.zpopmax('embedder:queue', 1)
+    if (maybeRedisQItem.length < 2) return // Q is empty, all done!
+
+    const [id, _] = maybeRedisQItem
+    if (!id) {
+      console.warn(`De-queued undefined item from embedder:queue`)
+      continue
+    }
+    const jobQueueId = parseInt(id, 10)
+    const jobQueueItem = await selectJobQueueItemById(jobQueueId)
+    if (!jobQueueItem) {
+      console.warn(`Unable to fetch EmbeddingsJobQueue.id = ${id}`)
+      continue
+    }
+
+    const metadata = await selectMetadataByJobQueueId(jobQueueId)
+    if (!metadata) {
+      await updateJobState(jobQueueId, 'failed', {
+        stateMessage: `unable to fetch metadata by EmbeddingsJobQueue.id = ${id}`
+      })
+      continue
+    }
+
+    let fullText = metadata?.embedText
+    try {
+      if (!fullText) {
+        fullText = await createEmbeddingTextFrom(jobQueueItem, dataLoader)
```

Comment:
> +1 I think we're doing this after filtering orgs by feature flag to minimize workload. 

What I was sketching out the architecture, I was having a difficult time making the discovery of items to add to the metadata table stateless. That created a separation between "indexing items to metadata" from "creating embeddings". I'm open to figuring out ways to make this process simpler, better as we evolve it.

> Alternatively, we can consider embedText to be a function of the model because the input token limit dictates what the text is. If we do this, then we can remove embedText from the metadata & just store it on the model table.

I thought it might be handy to keep a "maximum tokens" version of the embedText on the metadata table – that way when we add additional models (which feels likely to me), we don't need to re-generate the embedText